### PR TITLE
Add/meta for Firehose

### PR DESCRIPTION
**Pull request by millerf**
Works alongside D141162-code

Based on @gibrown 's idea here:  p3btAN-2D0-p2#comment-20298

## Proposed changes:

* Add metada to the post depending if:
  * It has paid-content block
  * It is paywalled

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Sync to your sandbox or docker
* Create a post paywalled post
* On your sandbox
```
 wp post meta list POST_ID --url=DOMAIN_URL
 ```
 Depending on your post content, you should see:
`_jetpack_memberships_contains_paywalled_content` empty or with `1` (not paywalled, paywalled)
`_jetpack_memberships_contains_paid_content` empty or with `1` (not contians paid-content block, does contain paid-content block)

* Values should change when updating the post, too.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php [hunk 1]
```diff
@@ -37,6 +40,34 @@ function register_block() {
...
+		'jetpack_sync_post_meta_whitelist',
+		function ( $allowed_meta ) {
+			return array_merge(
+				$allowed_meta,
+				array(
+					META_NAME_CONTAINS_PAID_CONTENT,
+				)
+			);
+		}
+	);
```
Comment (n3f) [inline projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php:68]: IIRC, we used to do this but added the fields to the default sync file.  Is this something we should do here too?

File: projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php [hunk 2]
```diff
@@ -111,3 +142,19 @@ function stripe_nudge( $checkout_url, $description, $button_text ) {
...
 }
+
+/**
+ * Add a meta to prevent publication on firehose, ES AI or Reader
+ *
+ * @param int     $post_id Post id.
+ * @param WP_Post $post Post being saved.
+ * @return void
+ */
+function maybe_add_paid_content_post_meta( int $post_id, WP_Post $post ) {
```
Comment (n3f) [inline projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php]: Doesn't seem accurately named: `maybe_add_paid_content_post_meta` ==> `add_paid_content_post_meta`

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php [hunk 3]
```diff
@@ -186,6 +202,25 @@ function register_newsletter_access_column( $columns ) {
 	);
 }
 
+/**
+ * Add a meta to prevent publication on firehose, ES AI or Reader
+ *
+ * @param int $post_id Post id being saved.
+ * @return void
+ */
+function maybe_add_paywalled_content_post_meta( int $post_id ) {
```
Comment (n3f) [inline projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: Again, this function adds meta no matter what so maybe rename?

**Review by n3f — DISMISSED** [review]
Works! I can approve, but I think you should probably rename the functions I mentioned.  Also not sure if the jetpack sync should be the final commit or just make it a stop-gap until the sync file gets updated (see code comment).

File: projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php (hunk 1 above)
Comment (n3f) [reply projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php:68]: Actually I think there's a dynamic whitelist for some prefixes 🤔.

Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php:68]: @n3f 

> but added the fields to the default sync file

I thought the same at first. But what we added in the default sync file are the whitelisted custom-post types for coupons and memberships (see [here](https://github.com/Automattic/jetpack/blob/1adad87a1c3d3ba419a7091786b98eb3153c5bd4/projects/packages/sync/src/class-defaults.php#L446)).  

You can see the other metas https://github.com/Automattic/jetpack/pull/36254/files#diff-9164d71f603706b3c6cc58c71ba3f2fbb12b1dba772adca76793f60c24ddbca7R143-R144

File: projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php (hunk 3 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/subscriptions/subscriptions.php]: Ah, true.

File: projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php (hunk 2 above)
Comment (millerf) [reply projects/plugins/jetpack/extensions/blocks/premium-content/premium-content.php]: Done.

**Review by n3f — APPROVED** [review]
LGTM

**Pull request merged**